include/unifex/async_scope.hpp

```diff
@@ -632,76 +625,129 @@ class _attached_op<Sender, Receiver>::type final {
   using cleaning_receiver_t = cleaning_receiver<type, Receiver>;
   using nested_operation_t = connect_result_t<Sender, cleaning_receiver_t>;
 
+  struct stop_callback {
+    void operator()() noexcept { op_.request_stop(); }
+
+    type& op_;
+  };
+
 public:
   template <typename Receiver2>
   explicit type(Sender&& s, Receiver2&& r, async_scope* scope) noexcept(
       is_nothrow_connectable_v<Sender, cleaning_receiver_t>&&
-          std::is_nothrow_constructible_v<cleaning_receiver_t, type*, Receiver2>&&
+          std::is_nothrow_constructible_v<cleaning_receiver_t, type*>&&
               std::is_nothrow_constructible_v<Receiver, Receiver2>)
-    : scope_(reinterpret_cast<std::uintptr_t>(scope) & started_mask)
-    , receiverToken_(get_stop_token(r)) {
-    if (scope_ref()) {
-      unifex::activate_union_member_with(op_, [&] {
-        return unifex::connect(
-            (Sender &&) s, cleaning_receiver_t{this, (Receiver2 &&) r});
+    : scope_(set_refcount(scope))
+    , receiver_((Receiver2 &&) r) {
+    if (scope) {
+      op_.construct_with([&, this] {
+        return unifex::connect((Sender &&) s, cleaning_receiver_t{this});
       });
-    } else {
-      unifex::activate_union_member(receiver_, (Receiver2 &&) r);
     }
   }
 
   type(type&& op) = delete;
 
   ~type() {
-    auto scope = scope_ref();
-    if (scope) {
-      unifex::deactivate_union_member(op_);
+    auto scope = scope_.load(std::memory_order_relaxed);
+    if (scope != 0u) {
+      op_.destruct();
       // started => receiver responsible for cleanup
-      if (!has_started()) {
-        record_done(scope);
+      if ((scope & ~mask) != 0u) {
+        UNIFEX_ASSERT((scope & ~mask) == 1u);
+        record_done(reinterpret_cast<async_scope*>(scope & mask));
       }
-    } else {
-      unifex::deactivate_union_member(receiver_);
     }
   }
 
-  auto get_token() { return stopSource_.get_token(); }
+  void request_stop() noexcept {
+    // increment from 0 means lost race with a completion method so
+    // no-op and let the in-flight completion complete
+    auto scope = scope_.load(std::memory_order_relaxed);
+    auto expected = (scope & mask) | 1u;
+    if (!scope_.compare_exchange_strong(
+            expected, expected + 1u, std::memory_order_relaxed)) {
+      return;
+    }
+    stopSource_.request_stop();
+    deliver_result([](auto&& r) noexcept { unifex::set_done(std::move(r)); });
+  }
+
+  // reduce size of the templated `deliver_result`
+  async_scope* deliver_result_prelude() noexcept {
+    auto scope = scope_.fetch_sub(1u, std::memory_order_acq_rel);
+    if ((scope & ~mask) != 1u) {
+      UNIFEX_ASSERT((scope & ~mask) == 2u);
+      return nullptr;
+    }
+    UNIFEX_ASSERT((scope & mask) != 0u);
+    deregister_callbacks();
+    return reinterpret_cast<async_scope*>(scope & mask);
```

Comment:
If you had the static helpers I described above, this could be:
```suggestion
    if (ref_count(scope) != 1u) {
      UNIFEX_ASSERT(ref_count(scope) == 2u);
      return nullptr;
    }
    UNIFEX_ASSERT(scope_ptr(scope) != nullptr);
    deregister_callbacks();
    return scope_ptr(scope);
```